### ISSUE TITLE
App Settings / Profiles have interference from Prompts, Metadata Update

### ISSUE THREAD

**Issue by dhaavi** [safing/portmaster#391]
**What happened**:

We've had multiple reports within the last months where settings were not saved, disappeared or were not added at all. These related issues may or may not be related and may or may not be fully resolved: #260, #312, #368, https://github.com/safing/portmaster-ui/issues/118

None of these had a particularly clean solution. I suspect that there is still some interference when saving profiles, or that profiles get "out of sync" so they are not saved and/or applied correctly.

**What did you expect to happen?**:

All settings should be preserved, always.

**How did you reproduce it?**:

We were not yet able to reproduce.

A user reported that this might be able to reproduce it:
1. use main app settings to: *allow domains* / *block domains*
2. use the left panel to: *allow domains* / *block domains*
3. Go back to the main app settings and all outgoing settings are wiped!

**Comment by northys** [comment]
All I can say is that the "didn't save" case never happened to me. It always disappears after a while. I whitelist analytics.google.com (which is blocked by ublock anyway) at least twice a month because I sometimes have to check the stats for client websites. Before I had troubles with accessing AWS console because they load CSS and JS from random cloudflare domain that portmaster do not like so I used whitelist for this domain but it always vanished after a while.

Please try the following - whitelist analytics.google.com (or any other domain) in this cute drop-down and give it a month at max. It works for a few days. I'll try to measure the time until it vanish. Experiment started right now.

![Screenshot from 2021-09-26 23-34-58](https://user-images.githubusercontent.com/2878126/134824934-2b00f3ff-5d8f-47d2-a641-43dff022a002.png)

**Comment by northys** [comment]
@dhaavi hi there after 22 hours.

```

**Comment by northys** [comment]
Interesting thing is that I have some domains allowed for a long time and I didn't have to allow them again. E.g. binance.com and the mentioned aws tracking code that broke aws console service search omnibox. analytics.google.com is gone though.

![Screenshot from 2021-09-29 15-57-23](https://user-images.githubusercontent.com/2878126/135283343-c11c1b8c-829b-4853-a145-9dd3205a14f8.png)

**Comment by ppacher** [comment]
Thanks for you investigation on that topic @northys! I'll try to reproduce the issue but since it does not affect all rules it's still seems somewhat random. I'll keep you updated!

**Comment by thegranddesign** [comment]
This effectively happens for me on all rules every time if there's something I can do to help debug this.  Obviously this bug makes portmaster effectively useless :joy:

**Comment by working-name** [comment]
Fresh user, before I got the idea that you can have an `Internet` rule to allow/deny, I had allowed and denied a bunch of domains for Chrome.exe. I then added Internet as the last rule, with allow. It works ok but seemingly randomly (within minutes however) it reverts the list of rules to what it was before I added the `Internet` rule. It happened several times but now that I switched to developer mode, of course, having trouble reproducing it.

**Comment by dhaavi** [comment]
> This effectively happens for me on all rules every time if there's something I can do to help debug this. Obviously this bug makes portmaster effectively useless

Hey @thegranddesign, I'm sorry to hear that this affects you so greatly. From what we could tell until now, it only happened here and there. Thanks for your input.

---

Hey @working-name, thanks for your report!

> I then added Internet as the last rule, with allow.

I'd recommend add Internet as a rule, as this will override other features, such as the filter lists.
Better just set the [Default Network Action](https://docs.safing.io/portmaster/settings#filter/defaultAction) to "Allow" and block LAN access if you want to.

> It works ok but seemingly randomly (within minutes however) it reverts the list of rules to what it was before I added the `Internet` rule.

This is exactly what we are investigating.

> Now new rules don't get saved.

This however is new. Very interesting.
Thanks also for the debug data, but unfortunately, there wasn't anything useful in there.

**Comment by working-name** [comment]
> Hey @working-name, thanks for your report!
> 
> > I then added Internet as the last rule, with allow.
> 
> I'd recommend add Internet as a rule, as this will override other features, such as the filter lists. Better just set the [Default Network Action](https://docs.safing.io/portmaster/settings#filter/defaultAction) to "Allow" and block LAN access if you want to.

Hi! I thought about that too, but for some apps the setting is not saved, it reverts back to "Ask".

> 
> > It works ok but seemingly randomly (within minutes however) it reverts the list of rules to what it was before I added the `Internet` rule.
> 
> This is exactly what we are investigating.
> 
> > Now new rules don't get saved.
> 
> This however is new. Very interesting. Thanks also for the debug data, but unfortunately, there wasn't anything useful in there.

How can I otherwise help?

I feel like I'm cluttering this thread since this issue is different. Should I open a new issue? Have some new odd behavior from this AM - existing rule is overwritten when selecting allow/deny.

**Comment by dhaavi** [comment]
> I thought about that too, but for some apps the setting is not saved, it reverts back to "Ask".

I see. Just retry in that case. ;) I believe the issue is some kind of race condition.

> How can I otherwise help?

Thanks for asking. Right now, I'll need to find the time to dig into this with the new information we have gathered. We are currently completing some big stuff for the SPN. I hope to find the time in a couple weeks to take a deep dive.

> I feel like I'm cluttering this thread since this issue is different. Should I open a new issue?

No - I'm pretty sure this is the exact issue.

**Comment by northys** [comment]
Right now I can't access my laptop though ssh because portmaster allow override for sshd have vanished. But yeah, focus on developing SPN so you can make more money on something that can't save value under a key persistently.

**Comment by working-name** [comment]
@northys It definitely is busy work setting the same thing over and over again and wondering why apps suddenly cannot connect. But in all honesty I'll take this over anything else I've tried. And I have tried pretty much everything including disgusting stuff like zonealarm and eset internet security.

Since you're on linux, you might want to give opensnitch a try. For windows folk everything out there is fairly trash. I miss macOS and little snitch or hands off!, but that OS is no better than windows nowadays with the amount of data leaks and collection and overriding power users like we're incapable of making sane decisions. Also mained fedora 34 but that ish is so buggy I wanted to throw my screen out the window. Speaking of, opensnitch has issues too, it'll sometimes "forget" that it's supposed to consult the list of rules and just denies every connection without logging said events - until you restart the daemon. That happens about 1-2 times per week. Or at least it used to, about 2-3 months ago when I was using it.

As far as prioritizing stuff that makes money - it is what it is. We all have bills and kids who want the new iPhone for Christmas and other such adulting oddities.

**Comment by davegson** [comment]
Hey northys,

we do not recommend using Portmaster / SPN for important things in your life. We clearly communicate that the software is in alpha and that you should expect bugs. 

We also [just acknowledged](https://safing.io/blog/2021/12/07/progress-update-november/#tablecard-in_progress-1) that we neglected bug fixing recently, but are re-shifting our attention to them. We know this is an important part of software development and we cannot wait to push Portmaster & SPN into Beta and 1.0.

I and we do appreciate all of your efforts. You are one of our biggest supporters and have been putting in loads of hours into testing and reporting ever since the very raw Portmaster days. 

Maybe because of that, issues like these can be extra frustrating; but being snarky about our priorities is neither helpful nor respectful. Instead, I would recommend re-evaluating the domains in which you use the Portmaster. And for some use cases opensnitch might be the better call for now as @working-name just recommended.

**Comment by northys** [comment]
Hi @davegson I dont need portmaster nor opensnitch at all. I'm just testing something which is IMHO a great piece of software. I just cant understand why some paid addon in something which has this sort of nasty bug higher priority for you. I would love to pay you few bucks to try it but why would I pay for something that once decides to deny me to connect to my computer? And why would anyone else would like to do that? I know this is OSS so I should STFU but I'm still confused about this. This is basically the first issue reported by multiple users and it makes portmaster unusable.

Its like hey we know portmaster is unusable because of a bug but you can use its premium feature that works perfectly!

The only place I have portmaster installed on is my laptop. I was just too lazy to get out of bed and I needed to send config for something to my friend. I wouldnt install it on my home server (which has gui actually lol).

**Comment by dhaavi** [comment]
From a technical side this bug is very challenging, as it is not about just storing a value in a database, but about multiple systems applying settings at the same time, ie. concurrency.
Especially, as I haven't been able to reproduce the bug.

I had a chance to look into this last week, which resulted in https://github.com/safing/portmaster/pull/461, to be released to the Beta Release Channel next week.
While I wasn't able to reproduce the issue, I have opted for numerous improvements to combat possible concurrency issues.
I am eager to see if that fixes or at least improves this problem.

**Comment by davegson** [comment]
> I would love to pay you few bucks to try it but why would I pay for something that once decides to deny me to connect to my computer? And why would anyone else would like to do that? I know this is OSS so I should STFU but I'm still confused about this. 

We all view the world through different lenses. Next to Daniel's technical input, I can give you some perspectives viewing through "a company lens". 

We always have an endless amount of work ahead of us. We currently have 78 open issues in this repo alone and have even more open work packages in our project management tool. What gets worked on is always a question of priority. 

---

And there are two main motivators why SPN got more priority:

1) From a _financial perspective_, it is Safing's goal to become independent as soon as possible. We are thankful for the grants we receive, but they have their cost too. It requires (A) _a lot_ of reporting and (B) puts restrictions on things like where we can hire people from. iE: We want another developer to work on Portmaster but being restricted to talent in Austria, our search has been challenging.

2) From a _product perspective_, it was long overdue that a public SPN version came out. And no, even this rather serious bug with its negative comes nowhere close to the positives of reaching this milestone. A lot of things start to change after this. Only after a product is "in the wild", a product can truly evolve and mature. The Portmaster went through the same process.

---

But all of this is rather off-topic. Fyi, we are also evaluating if we should activate GitHub discussions, which would have more room for these kind of conversations.

**Comment by dhaavi** [comment]
We now have multiple fixes that will hopefully resolve this issue in the Beta [Release Channel](https://docs.safing.io/portmaster/settings#core/releaseChannel).

**Comment by working-name** [comment]
Hi @dhaavi, tested the beta release and unfortunately I found the behavior to be similar, but somewhat worse. Difficult to put my finger on it, but I think it's safe to assume something goes very wrong when it tries to save the rules for an app - almost like a fail to write but the UI doesn't refresh so as a user you don't see what's going on.

There are other UI regressions but that's to be expected I assume, since it's a beta release.

https://www.dropbox.com/s/b5kpmuo9foy9ra6/2021-12-16%2003-45-37.mp4?dl=0 (view)

https://www.dropbox.com/s/b5kpmuo9foy9ra6/2021-12-16%2003-45-37.mp4?dl=1 (download directly)

**Comment by dhaavi** [comment]
Hey @working-name, thanks for testing this!

Also thanks for the video, that is great way to see what is going on!
I do believe that the rules are being saved, but the prompts are saved differently from the settings, and they seem to be interfering with each other.
I'm going to look into this.

**Comment by northys** [comment]
@dhaavi @working-name  isnt the problem the internal spotify resolver? i can see that allow button from widget clearly somehow overrides the manually set inputs but i think the whole problem with spotify at the beggining was its internal resolver and i think it somehow bypasses spotify. good job with the recording! i wasnt able to reproduce it so well.

all i can say is that i probably had the same issue with analytics / binance whitelisting because i probably enabled (im 70% sure) google analytics in the allow list manually but i definitely enabled binance using the connection history where you can toggle enable / disable for certain blocked domain.

the problem with ssh was different though. i havent opened portmaster dashboard for weeks when it happened. i had a 10.10.0.* manually written allow rule there and it somehow dissapeared without changing anything regarding ssh in the dashboard ui. as i said i havent opened it for weeks because everything worked smoothly so i blamed some kind of update.

**Comment by dhaavi** [comment]
Yes, I can confirm that this is a new bug introduced as part of the fixes in the previous release.
A fix is already in review and will be released today.

**Comment by dhaavi** [comment]
The fix is now released in v0.7.12 to the Beta Release Channel (as the broken version v0.7.11).

@working-name, please test this again. Everything should now work as expected.

As I get confirmation of this fix, I will mark the last couple comments as off-topic, as they were about a different - albeit very much related - bug than the original issue.

**Comment by working-name** [comment]
@dhaavi Awesome, thank you, I think that was the fix! Rules are maintained across manual edits and prompts, and restarts of portmaster. Including a video but ... it's kinda boring because everything works.

[Preview](https://www.dropbox.com/s/51w0v5reigbjja3/2021-12-19%2019-43-57.mp4?dl=0)
[Download](https://www.dropbox.com/s/51w0v5reigbjja3/2021-12-19%2019-43-57.mp4?dl=1)

Great job and thank you once again!

**Comment by thegranddesign** [comment]
@northys at the end of the day, I'm getting the feeling of a sense of entitlement from what you expect from a free and open source product.  I think in the future you're going to be much better off by coming from a perspective of assuming good intent rather than being snarky.

And to the Safing crew, I'm going to test this fix in a few days and I'll report back.  Thank you for the diligence here. :heart:   Also, I'll be _thrilled_ to pay for any open source product on an annual basis once it hits 1.0.  Linux is in desperate need of a good solution here.

**Comment by northys** [comment]
@thegranddesign I already apologized for being rude. It was stupid from me and I shouldn't be rude when something gets broken. It was just one of those bad mornings for me and I know that doesn't execuse my behaviour.

Sometimes I'm just too impulsive when I should rather stay quiet.

**Comment by thegranddesign** [comment]
@northys we've all been there!  Sorry I missed the apology otherwise I wouldn't have said anything further. :heart:

**Comment by northys** [comment]
@thegranddesign it's not here, I wrote a PM.

So once again on public I'm sorry I behaved like a jerk.

Now I see Daniel who seems to love everything organized spends another 30s marking this as off-topic :D

**Comment by dhaavi** [comment]
I found a possible regression during investigation of #484 - which itself might be a duplicate of this issue.

The first time during the handling of this whole issue I was able to reliably reproduce it for the first time: 🎉

- Set `curl`'s default network actions to "prompt".
- `curl -I example.org` waits for prompt.
- Allow prompt; command completes.
- A new rule was added: `+ .example.org.`
- `curl -I example.org` completes.
- `curl -I www.example.org` waits for prompt. 🐞
- Deny prompt; command fails.
- A new rule was added: `- .www.example.org.`
- Move `- .www.example.org.` below `+ .example.org.` rule.
- `curl -I example.org` completes.
- `curl -I www.example.org` completes as it matches `+ .example.org.`.

Will further investigate in the coming weeks.

**Comment by dhaavi** [comment]
Sorry, issue was closed by misconfigured bot.

**Comment by dhaavi** [comment]
I've just released a fix for the regression mentioned above in v0.8.2, available on the [Beta Release Channel](https://docs.safing.io/portmaster/settings#core/releaseChannel).

This might just be the last and we can finally close the chapter of all these related issues.

Whoever experienced an issue in this area, please update, test and report back.
